File: src/main/java/gg/manny/forums/Application.java
package gg.manny.forums;

import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import gg.manny.forums.rank.Rank;
import gg.manny.forums.rank.RankRepository;
import gg.manny.forums.web.controller.StaffController;
import lombok.Getter;
import org.commonmark.parser.Parser;
import org.commonmark.renderer.html.HtmlRenderer;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

import java.util.Arrays;
import java.util.Random;

@Getter
@SpringBootApplication
public class Application {

    public static final Random RANDOM = new Random();

    public static final Parser MARKDOWN_PARSER = Parser.builder().build();
    public static final HtmlRenderer MARKDOWN_RENDERER = HtmlRenderer.builder().build();

    public static final Gson GSON = new GsonBuilder()
            .serializeNulls()
            .setPrettyPrinting()
            .create();

    @Getter private static Application instance;

    public Application() {
    }

    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);

        instance = new Application();
    }

    public static void debug(String title, String... messages) {
        System.out.println(" - - - - - - - DEBUGGING: " + title + " - - - - - - - - - ");
        for (String message : messages) {
            System.out.println(" - -> " + message);
        }
        System.out.println(" - - - - - - - - - END OF DEBUG - - - - - - - - - - - ");
    }

    @Bean
    CommandLineRunner init(RankRepository roleRepository) {

        return args -> {
            if (!roleRepository.findById("default").isPresent()) {
                Rank role = new Rank();
                role.setId("default");
                role.setName("Default");
                role.setColor("&f");
                role.setForumColor("#ffff");
                role.setWeight(0);
                roleRepository.save(role);
            }
        };

    }
}

File: src/main/kotlin/dev/foraged/forums/Application.kt
package dev.foraged.forums

import com.google.gson.GsonBuilder
import com.google.gson.InstanceCreator
import com.minexd.core.CoreShared
import com.minexd.core.plugin.Plugin
import com.minexd.core.plugin.PluginEventHandler
import com.minexd.core.profile.Profile
import com.minexd.core.profile.ProfileService
import com.minexd.core.profile.grant.Grant
import com.minexd.core.profile.punishment.Punishment
import com.minexd.core.rank.Rank
import dev.foraged.commons.CommonsShared
import dev.foraged.forums.leaderboard.LeaderboardService
import dev.foraged.forums.profile.SiteProfile
import dev.foraged.forums.shop.ShopMessages
import dev.foraged.shop.ShopShared
import gg.scala.aware.AwareHub
import gg.scala.aware.uri.WrappedAwareUri
import gg.scala.store.ScalaDataStoreShared
import gg.scala.store.connection.mongo.AbstractDataStoreMongoConnection
import gg.scala.store.connection.mongo.impl.UriDataStoreMongoConnection
import gg.scala.store.connection.mongo.impl.details.DataStoreMongoConnectionDetails
import gg.scala.store.connection.redis.AbstractDataStoreRedisConnection
import gg.scala.store.connection.redis.impl.DataStoreRedisConnection
import gg.scala.store.serializer.serializers.GsonSerializer
import net.evilblock.cubed.serializers.Serializers
import org.commonmark.parser.Parser
import org.commonmark.renderer.html.HtmlRenderer
import org.springframework.boot.CommandLineRunner
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.ComponentScan
import revxrsal.commands.cli.ConsoleCommandHandler
import java.lang.reflect.Type
import java.util.*
import java.util.logging.Logger

@SpringBootApplication
@ComponentScan("dev.foraged.forums")
open class Application : ScalaDataStoreShared(), Plugin, PluginEventHandler
{
    override fun getProfileType(): Type { return Profile::class.java }
    override fun getLogger(): Logger { return Logger.getAnonymousLogger() }
    override fun createProfileInstance(uuid: UUID) = SiteProfile(uuid)
    override fun getEventHandler(): PluginEventHandler { return this }
    override fun getActiveGroups(): Set<String> { return setOf("GLOBAL") }
    override fun callGrantUpdateEvent(target: Profile, grant: Grant)
    {
        TODO("Not yet implemented")
    }

    override fun callPunishmentUpdateEvent(target: Profile, punishment: Punishment)
    {
        TODO("Not yet implemented")
    }

    override fun callRankUpdateEvent(rank: Rank)
    {
        TODO("Not yet implemented")
    }

    override fun getNewRedisConnection(): AbstractDataStoreRedisConnection {
        return DataStoreRedisConnection()
    }

    override fun debug(from: String, message: String) {
        getLogger().info("[${from}] [debug] $message")
    }

    override fun getNewMongoConnection(): AbstractDataStoreMongoConnection {
        return UriDataStoreMongoConnection(DataStoreMongoConnectionDetails("mongodb://localhost:27017/Scala", "Scala"))
    }

    companion object {
        val COMMAND_HANDLER = ConsoleCommandHandler.create()
        val MARKDOWN_PARSER = Parser.builder().build()
        val MARKDOWN_RENDERER = HtmlRenderer.builder().build()
        val GSON = GsonBuilder()
            .serializeNulls()
            .create()

        lateinit var INSTANCE: Application
        lateinit var CONTEXT: ConfigurableApplicationContext

        @JvmStatic
        fun main(args: Array<String>) {
            CONTEXT = SpringApplication.run(Application::class.java, *args)
            INSTANCE = Application()
            AwareHub.configure(WrappedAwareUri()) {
                GSON!!
            }
            ScalaDataStoreShared.INSTANCE = INSTANCE
            Serializers.useGsonBuilderThenRebuild {
                it.registerTypeAdapter(Profile::class.java, object : InstanceCreator<Profile> {
                    override fun createInstance(p0: Type): Profile {
                        return SiteProfile(UUID.randomUUID())
                    }

                }).create()
            }
            GsonSerializer.provideCustomGson {
                Serializers.gson
            }
            CoreShared(INSTANCE).configure()
            CommonsShared.configure()
            ShopShared().configure(CONTEXT.getBean(ShopMessages::class.java))
            LeaderboardService.configure()

            COMMAND_HANDLER.pollInput()
        }
    }
}